server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -462,9 +464,10 @@ public final void onShardFailure(final int shardIndex, @Nullable SearchShardTarg
      * @param result the result returned form the shard
      * @param shardIt the shard iterator
      */
-    private void onShardResult(Result result, SearchShardIterator shardIt) {
+    protected void onShardResult(Result result, SearchShardIterator shardIt) {
         assert result.getShardIndex() != -1 : "shard index is not set";
         assert result.getSearchShardTarget() != null : "search shard target must not be null";
+        hasShardResponse.set(true);
```

Comment:
nit: can we use `successfulOps` instead? Also, I think we can have up to `maxConcurrentRequestsPerNode` responses instead of one. But I don't see a way to avoid this as any shard request can be failed.

Reply:
Ah, I missed that. Thanks for the explanation.